server/auth.go

```diff
@@ -557,6 +578,54 @@ func (s *Server) processClientOrLeafAuthentication(c *client, opts *Options) boo
 		err  error
 		ao   bool // auth override
 	)
+
+	// Check if we have auth callouts enabled at the server level or in the bound account.
+	defer func() {
+		if juc == nil && opts.AuthCallout == nil {
+			return
+		}
+		// We have a juc define here, check account.
+		if juc != nil && !acc.hasExternalAuth() {
+			return
+		}
+
+		// We have auth callout set here.
+		var skip bool
+		if authorized {
+			// Check if we are on the list of auth_users.
+			userID := c.getRawAuthUser()
+			if juc != nil {
+				skip = acc.isExternalAuthUser(userID)
+			} else {
+				for _, u := range opts.AuthCallout.AuthUsers {
+					if userID == u {
+						skip = true
+						break
+					}
+				}
+			}
+		}
+		// If we are here we have an auth callout defined and we have failed auth so far
+		// so we will callout to our auth backend for processing.
+		if !skip {
+			authorized = s.processClientOrLeafCallout(c, opts)
+		}
+		// Check if we are and in the auth callout account, and if so add in deny publish permissions for the auth subject.
+		if authorized {
+			var authAccountName string
+			if juc == nil && opts.AuthCallout != nil {
+				authAccountName = opts.AuthCallout.Account
+			} else if juc != nil {
+				authAccountName = acc.Name
+			}
+			c.mu.Lock()
+			if c.acc != nil && c.acc.Name == authAccountName {
+				c.mergeDenyPermissions(pub, []string{AuthCalloutSubject})
+			}
+			c.mu.Unlock()
```

Comment:
If this weren't inside a huge defer doing critical work, then you could make this a more normal/safer deferred call to unlock.

Reply:
625 matches the lock on 621. Normal locking unlocking. For small blocks without branching I enclose, for more complicated stuff I try to leverge defer unlock. Defer could be used in hear but is function scoped, not block.